add infrastructure for binary LGPL compliance checks

kaspar030 opened the pull request:
generalized version of #2357.
Work in progress. beauty-patches welcome! README.md outdated.

This PR adds some new make targets:
```patch``` -> apply all patches in patches/ folder to $RIOTBASE using ```git am```
```bindist``` -> creates a folder "bindist" which contains all files needed for LGPL-compliant distribution
```check_bindist``` -> supposed to be run from within a folder created by "make bindist". Checks if compiled RIOTBASE linked with included .a files md5-matches included .elf.

Please check out the concept before I thorougly document it. ;)

emmanuelsearch commented:
@kaspar030 so where do we stand on this one?

Kijewski commented on Makefile.base:
```diff
@@ -16,8 +16,9 @@ all: $(BINDIR)$(MODULE).a ..nothing
+# Build subdirs. Ignore error if DIR is in WEAK_DIRS.
 ${DIRS:%=ALL--%}:
-	"$(MAKE)" -C ${@:ALL--%=%}
+	"$(MAKE)" -C ${@:ALL--%=%} || [[ "$(WEAK_DIRS)" != *${@:ALL--%=%}* ]]
```
The test is in the wrong order, isn't it?

OlegHahm commented:
@kaspar030 so where do we stand on this one?

OlegHahm commented:
@kaspar030?

OlegHahm commented:
:beer: ?

kaspar030 commented:
I rebased it, but the make targets don't work. I'll have no time until monday to investigate.
And, this is still RFC in terms of concept, noone commented...

emmanuelsearch commented:
Would be awesome to have some progress on this...

kaspar030 commented:
- rebased
- dropped library example. @BytesGalore, it seemed too complicated, and I'm not even sure it's concept doesn't make the whole a derived work.

emmanuelsearch commented:
So is this still WIP, or not?

emmanuelsearch commented:
Does the "example" code here not belong more in "dist/tools/" ?

Kijewski commented:
Maybe $RANDOM_PERSON could have a look how the Debian folks implemented their reproducible builds: https://wiki.debian.org/ReproducibleBuilds

emmanuelsearch commented:
I think this new feature is great!

emmanuelsearch commented:
Tested. I get this error. Is there a file missing in the commits?
```
RIOT/Makefile.bindist:4: Makefile.distcheck: No such file or directory
make: *** No rule to make target 'Makefile.distcheck'.  Stop.
```

emmanuelsearch commented:
A separate OS X issue: on Mac I get this error at when I make bindist:

```
In file included from /RIOT/cpu/native/periph/timer.c:42:
In file included from /RIOT/cpu/native/include/native_internal.h:177:
/RIOT/core/include/sched.h:97:24: error: typedef
      redefinition with different types ('struct _thread' vs 'mach_port_t' (aka 'unsigned int'))
typedef struct _thread thread_t;
                       ^
/usr/include/mach/mach_types.h:119:22: note: previous definition is here
typedef mach_port_t             thread_t;
                                ^
1 error generated.

```

BytesGalore commented:
@kaspar030 I guess you're right, the library linking example is something ontop of the base functionality of this PR.
I still think this PR is very valuable to help/enable easy checking of LGPL compliance for shipped binary blobs.

emmanuelsearch commented:
@kaspar030 for some reason it keeps telling me ```Warning! RIOT_VERSION doesn't match!``` and the .elf don't match even though I am using the same commit and the same toolchain, when I ```make check_bindist```
I suppose I need to patch at some point, and that's the reason? But what, how and when?

kaspar030 commented:
Hm, ```RIOT_VERSION doesn't match``` is weird.

Can you try adding ```RIOT_VERSION_OVERRIDE:=test``` to the makefile, remove bindist, then ```make clean all``` again?

emmanuelsearch commented:
Tested, works even without the RIOT_VERSION_OVERRIDE:=test
I think the README should be modified to explain that the process is:

1. make all
2. make bindist
3. cd bindist
4. make check_bindist

Is there also a way to automate that the RIOTBASE in the bindist/Makefile automatically appends a /.. by default? (if a user needs to update it nevertheless, it does not change anything, else it saves one manual edit ;)

emmanuelsearch commented:
@kaspar030 so what's the latest on this? Any chance this makes the feature freeze?

kaspar030 commented:
> I think the README should be modified to explain that the process is:

updated.

> Is there also a way to automate that the RIOTBASE in the bindist/Makefile automatically appends a /.. by default?

Hm, in principle, but other than make the example work, it would be fiddling with a user's makefile. I've added the step to the README, do you think this is enough?

> so what's the latest on this? Any chance this makes the feature freeze?

IMHO we can merge, if you give it an ACK. ;)

emmanuelsearch commented:
ACK!

kaspar030 commented:
- squashed, rebased, dropped the unneeded patch target commit, waiting for murdock

OlegHahm commented:
lol:
```
file has an unknown license header: 'examples/bindist/abc/abc.c'
```
I'm fine with merging this anyway. ;)

emmanuelsearch commented:
@kaspar030 can you add license + copyright P. Schmerzl on abc.c? ;)
should be good to go
